ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs

```diff
@@ -427,61 +572,87 @@ appendInputs ::
-        InputEBB{}    -> inp : go n inps
-        InputSigner{}
-          | n == 0    -> []
-          | otherwise -> inp : go (n - 1) inps
-
--- | May include EBBs after the last signer *BUT* *NOT* before the first signer
-takeLastSigners :: Word64 -> Inputs c -> Inputs c
-takeLastSigners = \n0 -> Inputs . reverse . go n0 . reverse . unInputs
+    prjSigner :: Input c -> Maybe (CS.PBftSigner c)
+    prjSigner = \case
+      InputEBB{}    -> Nothing
+      InputSigner x -> Just x
+
+spanEBBs :: Inputs c -> ([PBftEBB], Inputs c)
+spanEBBs (Inputs inps0) =
+    (pre, post)
   where
-    go 0 = const []
-    go n = \case
-      []       -> []
-      inp:inps -> inp : go n' inps
-        where
-          n' = case inp of
-            InputEBB{}    -> n
-            InputSigner{} -> n - 1
-
--- | Wrapper around 'CS.fromList' that also sets the 'ebbs' field
+    (pre, mbPost) = spanLeft prj inps0
+      where
+        prj = \case
+          InputEBB x        -> Left x
+          inp@InputSigner{} -> Right inp
+
+    post = case mbPost of
+      Nothing          -> mempty
+      Just (inp, inps) -> Inputs $ inp : inps
+
+-- | Wrapper around 'CS.fromList'
 fromInputs ::
      PBftCrypto c
   => SecurityParam
   -> CS.WindowSize
-  -> (WithOrigin SlotNo, Inputs c)
+  -> WithOrigin SlotNo
+  -> [CS.PBftSigner c]
+     -- ^ determines the window of signers
+  -> [PBftEBB]
+     -- ^ determines 'CS.ebbs'
   -> CS.PBftChainState c
-fromInputs k n (anchor, inputs) =
-    CS.fromList k n (anchor, Seq.fromList $ signatureInputs inputs, EbbMap m)
+fromInputs k n anchor signers ebbs =
+    CS.fromList k n (anchor, Seq.fromList signers, EbbMap m)
   where
-    m = Map.fromList [ (slot, mSlot) | (mSlot, slot) <- ebbInputs inputs ]
+    m = Map.fromList [ (slot, mSlot) | PBftEBB mSlot slot <- ebbs ]
 
-splitAtLastInputs :: Word64 -> Inputs c -> (Inputs c, Inputs c)
-splitAtLastInputs n (Inputs inps) = (Inputs l, Inputs r)
-  where
-    (l, r) = splitAt (length inps - fromIntegral n) inps
+{-------------------------------------------------------------------------------
+  "The previous input"
+-------------------------------------------------------------------------------}
 
-dropLastInputs :: Word64 -> Inputs c -> Inputs c
-dropLastInputs n = Inputs . dropLast (fromIntegral n) . unInputs
+-- | Summary of an 'Inputs' /that/ /begins/ /at/ 'Origin'
+newtype LatestInput c = LatestInput {unLatestInput :: Maybe (Input c)}
 
-dropWhileNotAfterInputs :: WithOrigin SlotNo -> Inputs c -> Inputs c
-dropWhileNotAfterInputs mSlot =
-  Inputs . dropWhile ((<= mSlot) . At . slotInput) . unInputs
+lastInput :: Inputs c -> LatestInput c
+lastInput = LatestInput . safeLast . unInputs
 
-takeLastInputs :: Word64 -> Inputs c -> Inputs c
-takeLastInputs n = Inputs . takeLast (fromIntegral n) . unInputs
+-- | The slot of the latest block
+slotLatestInput :: LatestInput c -> WithOrigin SlotNo
+slotLatestInput (LatestInput mbInp) = maybe Origin (At . slotInput) mbInp
```

Comment:
Would `Ouroboros.Network.Point.withOriginFromMaybe` make this clearer/

Reply:
Not in my opinion?

I changed it to

```
-- | The slot of the latest block
slotLatestInput :: LatestInput c -> WithOrigin SlotNo
slotLatestInput (LatestInput mbInp) = case mbInp of
  Nothing  -> Origin
  Just inp -> At $ slotInput inp
```